src/EventTypeDefs.ts

```diff
@@ -0,0 +1,228 @@
+  e: E;
+};
+
+export type BasicTransformEvent<E extends Event = TPointerEvent> = TEvent<E> & {
+  transform: Transform;
+  pointer: Point;
+};
+
+export type TModificationEvents =
+  | 'moving'
+  | 'scaling'
+  | 'rotating'
+  | 'skewing'
+  | 'resizing';
+
+type ObjectModifiedEvents = Record<TModificationEvents, BasicTransformEvent> & {
+  modified: BasicTransformEvent | never;
+};
+
+type CanvasModifiedEvents = Record<
+  `object:${keyof ObjectModifiedEvents}`,
+  BasicTransformEvent & { target: FabricObject }
+>;
+
+export type TransformEvent<T extends Event = TPointerEvent> =
+  BasicTransformEvent<T> & {
+    target: FabricObject;
+    subTargets: FabricObject[];
+    button: number;
+    isClick: boolean;
+    pointer: Point;
+    absolutePointer: Point;
+  };
+
+type SimpleEventHandler<T extends Event = TPointerEvent> = TEvent<T> & {
+  target: FabricObject;
+  subTargets: FabricObject[];
+};
+
+type InEvent = {
+  previousTarget?: FabricObject;
+};
+
+type OutEvent = {
+  nextTarget?: FabricObject;
+};
+
+type DragEventData = TEvent<DragEvent> & {
+  target: FabricObject;
+  subTargets?: FabricObject[];
+  dragSource?: FabricObject;
+  canDrop?: boolean;
+  dropTarget?: FabricObject;
+};
+
+type DropEventData = DragEventData & { pointer: Point };
+
+type DnDEvents = {
+  dragstart: TEvent<DragEvent> & { target: FabricObject };
+  drag: DragEventData;
+  dragover: DragEventData;
+  dragenter: DragEventData & InEvent;
+  dragleave: DragEventData & OutEvent;
+  dragend: DragEventData;
+  'drop:before': DropEventData;
+  drop: DropEventData;
+  'drop:after': DropEventData;
+};
+
+type CanvasDnDEvents = DnDEvents & {
+  'drag:enter': DragEventData & InEvent;
+  'drag:leave': DragEventData & OutEvent;
+};
+
+type CanvasSelectionEvents = {
+  'selection:created': TEvent & {
+    selected: FabricObject[];
+  };
+  'selection:updated': TEvent & {
+    selected: FabricObject[];
+    deselected: FabricObject[];
+  };
+  'before:selection:cleared': Partial<TEvent> & {
+    deselected: FabricObject[];
+  };
+  'selection:cleared': Partial<TEvent> & {
+    deselected: FabricObject[];
+  };
+};
+
+type BeforeSuffix<T extends string> = `${T}:before`;
+type WithBeforeSuffix<T extends string> = T | BeforeSuffix<T>;
+
+type TPointerEvents<Prefix extends string, E = Record<string, never>> = Record<
+  `${Prefix}${
+    | WithBeforeSuffix<'down'>
+    | WithBeforeSuffix<'move'>
+    | WithBeforeSuffix<'up'>
+    | 'dblclick'}`,
+  TransformEvent & E
+> &
+  Record<`${Prefix}wheel`, TransformEvent<WheelEvent> & E> &
+  Record<`${Prefix}over`, TransformEvent & InEvent & E> &
+  Record<`${Prefix}out`, TransformEvent & OutEvent & E>;
+
+export type ObjectPointerEvents = TPointerEvents<'mouse'>;
+export type CanvasPointerEvents = TPointerEvents<'mouse:'>;
+
+export type ObjectEvents = ObjectPointerEvents &
+  DnDEvents &
+  ObjectModifiedEvents & {
+    // selection
+    selected: never;
+    deselected: never;
```

Comment:
to me looks like that selected and deselected do get `target` and optionally 'e' ( if they happen by mouse )

Reply:
I recall that the object is fired without context
But need to double check

Reply:
As opposed to the events fired on canvas

Reply:
I verified and added e + target.